docs/design-documents/20231117-better-processors.md

```diff
@@ -0,0 +1,889 @@
+
+We will also need to provide an SDK for users who want to implement their own processor.
+Our proposal is to again mimic the structure we use for connectors and create a separate
+module (e.g. `github.com/conduitio/conduit-processor-sdk`) that exposes a simple SDK for
+implementing a processor plugin. This SDK should provide utilities for logging and
+calling exposed host functions. An example of such an SDK is the
+[Redpanda transform SDK](https://github.com/redpanda-data/redpanda/tree/dev/src/transform-sdk/go).
+
+### Conditional execution of processors
+
+Currently the only way to control what record gets passed to the processor is by
+attaching the processor either to a source or a destination connector. It would
+be useful if the user could additionally define a condition that must be fulfilled for
+the record to get passed to a processor.
+
+We already have a similar functionality in the `filterfieldkey` processor which lets you
+filter a record if some condition is met. The idea is to extract this functionality and
+create a generalized way to create conditions on processors. If the condition evaluates
+to `false`, the processor is skipped and the record is passed along to the next node
+without being processed.
+
+This way we can simplify the `filter` processor and, more importantly, give the user
+the control of defining when a processor should be executed.
+
+We propose adding a field `condition` in the processor configuration that allows you to
+write a Go template that returns a boolean. If the returned value evaluates to `true`,
+the processor will be executed, otherwise it will be skipped.
+If the condition field is empty it is the same as setting the condition to `true`.
+In other words, by default all records are passed to the processor.
+
+```yaml
+processors:
+  id:   example
+  type: custom.javascript
+  # Only execute processor if the metadata field "opencdc.version" contains value "v1"
+  condition: {{ eq .Metadata.opencdc.version "v1" }}
+  settings:
+    script: >
+      function process(record) {
+        logger.Info().Msg("executing JavaScript processor on OpenCDC v1 record");
+        return record;
+      }
+```
+
+### Addressing record fields
+
+Anytime a processor contains a setting that allows you to pick a field or write a
+condition we recommend using the same syntax as the Go stdlib package
+[text/template](https://pkg.go.dev/text/template). This way we are leaning on syntax
+that is thoroughly documented and has a lot of resources online outside of the
+Conduit ecosystem. Additionally, we already use the same approach in specifying the
+record [output format](https://conduit.io/docs/connectors/output-formats).
+
+Here are a few examples of addressing record fields:
+
+- `.` - references the whole record
+- `.Payload.After` - references the field that normally carries the payload
+- `{{ eq .Key.foo "123" }}` - evaluates to `true` if the key contains the value
```

Comment:
I think `eq` evaluates `==`, so it's true if they're equal not if it contains the value